### DISCUSSION TITLE
Elapsed time to download content is zero.

### DISCUSSION THREAD

**Discussion by BabakAmini** [aio-libs/aiohttp#8054] [Q&A] [answered]
As far as I know, `session.get()` loads headers from the destination server, and then an API like `resp.text()` downloads the entire content from it. I need to save elapsed time after each step for further investigation. Below is a sample code to do this task:

```Python
def do_request(self):
    self.headers_fetch_elapsed_time, self.text_fetch_elapsed_time = 0, 0
    t1 = time.perf_counter()

    async with self.session.get(self.url) as resp:
        t2 = time.perf_counter()
        self.headers_fetch_elapsed_time = round(1000 * (t2 - t1))
        self.txt = await resp.text()
        self.text_fetch_elapsed_time = round(1000 * (time.perf_counter() - t2))
```

I use the above function multiple times by a specific URL but most of the time `self.text_fetch_elapsed_time` value is zero. At first, I thought there might be a caching system in aiohttp but, I'm pretty sure that each request to the server returns a completely new response.

So, why is this happening?

**Comment by Dreamsorcerer** [comment, marked as answer]
> As far as I know, `session.get()` loads headers from the destination server, and then an API like `resp.text()` downloads the entire content from it.

Not exactly, that's not how HTTP works. It downloads a small amount of data (maybe 4KB or something) and parses the headers from that before returning. When you call any variation of .read() it will then continue from the buffered data and fetch anything more from the server if needed. So, if your responses are small, the network request will have been completed and the connection released before your call to .text().

**Reply by BabakAmini** [reply]
Thank you, @Dreamsorcerer, your point is clear and logical. For others who came across here, below is a small snippet of data I've saved that confirms what you said (content_length  and times are in bytes and milliseconds respectively):

![image](https://github.com/aio-libs/aiohttp/assets/85762931/b2af7e4b-f164-45f5-a962-b9dfd11fcc6d)

Also, now I know why `resp.connection` is not available in such a case. It becomes **None** after all data is downloaded and this becomes another problem I have.

I need to have access to the `resp.connection` attributes after the data is downloaded so, I save it immediately after `session.get ()`. But since sometimes data is fully downloaded, `resp.connection` will be **None** based on what you described. Is there a way to stop releasing the connection after the response is received and do it manually?

For your information, as the server rejects keep-alive connections, I have to use a TCPConnector with `force_close=True`.

**Reply by Dreamsorcerer** [reply]
> For your information, as the server rejects keep-alive connections, I have to use a TCPConnector with force_close=True.

Not that I'm aware of. Why would you want to access the closed (or possibly reused) connection? I can't imagine a case for general app development, but for internal aiohttp tests we've sometimes checked the connections directly in the connection pool.

**Reply by BabakAmini** [reply]
I need to save remote IPs for further investigations.

**Reply by Dreamsorcerer** [reply]
I'm not too familiar with it myself, but I think you can probably get that information through the tracing API (e.g. when DNS resolution happens or the connection is acquired): https://docs.aiohttp.org/en/stable/client_advanced.html#client-tracing

**Reply by BabakAmini** [reply]
Oh, I already implemented that part of my code but forgot about it. Maybe I should check it again and find something relative, it slows down the code though.

Thank you again for your assistance.

**Reply by Dreamsorcerer** [reply]
Well, without tracing, you could probably dig into the DNS cache and just pick out the IPs used there. I don't remember much about it, so you might need to dig into the code to figure that out.
